### PR TITLE
Reduce allocation for `GraphQL::Metrics.time`

### PR DESCRIPTION
**Pull request by catlee**
Using `*args` for `GraphQL::Metrics.time` seems to produce extra allocations, and is slower than accepting a single optional argument.

```
ruby 3.2.2 (2023-09-23 revision 6e984f9449) +YJIT [x86_64-linux]
=== Call with default offset_time ===
Warming up --------------------------------------
        Metrics.time   362.921k i/100ms
Metrics.time_without_splat
                       599.863k i/100ms
Calculating -------------------------------------
        Metrics.time      3.762M (± 8.5%) i/s -     18.872M in   5.070886s
Metrics.time_without_splat
                          6.440M (± 4.4%) i/s -     32.393M in   5.041691s

Comparison:
Metrics.time_without_splat:  6439740.1 i/s
        Metrics.time:  3761877.2 i/s - 1.71x  slower

Metrics.time: 2.002 alloc/iter
Metrics.time_without_splat: 1.001 alloc/iter

=== Call with explicit offset_time ===
Warming up --------------------------------------
        Metrics.time   321.524k i/100ms
Metrics.time_without_splat
                       576.466k i/100ms
Calculating -------------------------------------
        Metrics.time      3.404M (± 7.0%) i/s -     17.041M in   5.035086s
Metrics.time_without_splat
                          6.275M (± 6.2%) i/s -     31.706M in   5.077900s

Comparison:
Metrics.time_without_splat:  6274733.3 i/s
        Metrics.time:  3403527.6 i/s - 1.84x  slower

Metrics.time: 2.001 alloc/iter
Metrics.time_without_splat: 1.001 alloc/iter
```

```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'benchmark-ips'
end

def measure_allocs(title, iterations: 1_000)
  before = GC.stat(:total_allocated_objects)
  iterations.times do
    yield
  end
  allocs = GC.stat(:total_allocated_objects) - before
  puts "#{title}: #{allocs.to_f / iterations} alloc/iter"
end

puts RUBY_DESCRIPTION

class TimedResult
  attr_reader :offset_time, :result

  def initialize(offset_time = nil)
    @offset_time = offset_time
    @result = yield
  end
end

class Metrics
  def self.time(*args)
    TimedResult.new(*args) { yield }
  end

  def self.time_without_splat(offset_time = nil)
    TimedResult.new(offset_time) { yield }
  end
end

puts "=== Call with default offset_time ==="

Benchmark.ips do |x|
  x.report("Metrics.time") { Metrics.time { 1 + 1 } }
  x.report("Metrics.time_without_splat") { Metrics.time_without_splat { 1 + 1 } }
  x.compare!
end

measure_allocs("Metrics.time") { Metrics.time { 1 + 1 } }
measure_allocs("Metrics.time_without_splat") { Metrics.time_without_splat { 1 + 1 } }

puts
puts "=== Call with explicit offset_time ==="

Benchmark.ips do |x|
  x.report("Metrics.time") { Metrics.time(0) { 1 + 1 } }
  x.report("Metrics.time_without_splat") { Metrics.time_without_splat(0) { 1 + 1 } }
  x.compare!
end

measure_allocs("Metrics.time") { Metrics.time(0) { 1 + 1 } }
measure_allocs("Metrics.time_without_splat") { Metrics.time_without_splat(0) { 1 + 1 } }
```

(credit to @byroot for his allocation benchmark script [here](https://github.com/rails/rails/pull/50268))

### CODE REVIEW

File: lib/graphql/metrics.rb [hunk 1]
```diff
@@ -51,8 +51,8 @@ def self.current_time_monotonic
       Process.clock_gettime(Process::CLOCK_MONOTONIC)
     end
 
-    def self.time(*args)
-      TimedResult.new(*args) { yield }
+    def self.time(offset_time = nil)
+      TimedResult.new(offset_time) { yield }
```
Comment (byroot) [inline lib/graphql/metrics.rb:55]:
```suggestion
    def self.time(offset_time = nil, &block)
      TimedResult.new(offset_time, &block)
```

Comment (catlee) [reply lib/graphql/metrics.rb:55]: default argument:
```
Comparison:
Metrics.time_without_splat_and_block:  6667246.6 i/s
Metrics.time_without_splat:  6524671.1 i/s - same-ish: difference falls within error
        Metrics.time:  3756359.7 i/s - 1.77x  slower
```

explicit argument:
```
Comparison:
Metrics.time_without_splat:  6364099.5 i/s
Metrics.time_without_splat_and_block:  5018904.9 i/s - 1.27x  slower
        Metrics.time:  3401205.2 i/s - 1.87x  slower

```

Comment (byroot) [reply lib/graphql/metrics.rb:55]: How so? It's avoiding an extra block. It's not a big deal, but it's one less call.

Comment (byroot) [reply lib/graphql/metrics.rb:55]: I don't have the same results. Not sure which ruby you are using. But anyway, based on the implementation I'd be very surprised if `{ yield }` could be faster in any cases.

**Pull request merged**